bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -2410,19 +2304,24 @@ private static boolean isByteLiteral(long longValue) {
 
     private static void execIteratorOperation(WorkerExecutionContext ctx, WorkerData sf, Instruction instruction) {
         int i, j;
-        BCollection collection;
+        BValue collection;
         BIterator iterator;
         InstructionIteratorNext nextInstruction;
         switch (instruction.getOpcode()) {
             case InstructionCodes.ITR_NEW:
                 i = instruction.getOperands()[0];   // collection
                 j = instruction.getOperands()[1];   // iterator variable (ref) index.
-                collection = (BCollection) sf.refRegs[i];
+
+                collection = sf.refRegs[i];
                 if (collection == null) {
                     handleNullRefError(ctx);
                     return;
+                } else if (!(collection instanceof BCollection)) {
+                    sf.refRegs[j] = null;
```

Comment:
Instead of setting null, we might need to set to return an empty iterator with hasNext return false.